SPARKC-283: Adds CassandraOption for Skipping Empty Columns

RussellSpitzer opened the pull request:
Previously unset columns had to be set to null since the Cassandra
Protocol <= V3 did not support unbound parameters in bound statements.
The newere version of the protocol allows for this, and this patch
enables the connector to take advtange of that behavior.

This is done through use of a new option container, CassandraOption.
This class holds an Option and the behavior that a None should have upon
being written. With this class a row can have varying behavior from
column to column and row to row.

If an Unset is used on a C* version prior to Protcol Version 3 a warning
will be logged that the Unset functionality cannot be used and the value
will be treated as a Delete instead.

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

RussellSpitzer commented:
@jacek-lewandowski Rebased

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

jacek-lewandowski commented on spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/CassandraRDDSpec.scala:
```diff
@@ -332,6 +331,11 @@ class CassandraRDDSpec extends SparkCassandraITFlatSpecBase {
     rowById(2).getBytesOption("b") shouldEqual None
   }
 
+  it should "allow for reading Cassandra Options from nulls" in {
```
What about testing whether {{CassandraOption}} can be read from non-nulls

jacek-lewandowski commented on spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/writer/BoundStatementBuilderSpec.scala:
```diff
@@ -0,0 +1,39 @@
+  val protocolVersion = conn.withClusterDo(cluster => cluster.getConfiguration.getProtocolOptions.getProtocolVersion)
+
+  "BoundStatementBuilder" should "ignore Unset values if ProtocolVersion > 3" in {
+    val bsb = new BoundStatementBuilder(rowWriter, ps, protocolVersion = ProtocolVersion.V4)
```
perhaps should be iterated through protocol versions which `protocol.asInt >= ProtocolVersion.V4.asInt`

jacek-lewandowski commented on spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/writer/BoundStatementBuilderSpec.scala:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  it should "set Unset values to null if ProtocolVersion <= 3" in {
+    val bsb = new BoundStatementBuilder(rowWriter, ps, protocolVersion = ProtocolVersion.V3)
```
same here

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/ReadConf.scala:
```diff
@@ -6,7 +6,7 @@ import org.apache.spark.SparkConf
 /** Read settings for RDD
   *
-  * @param splitCount number of partitions to divide the data into; unset by default
+  * @param splitCount number of partitions to divide the data into; unsetIfNone by default
```
?

RussellSpitzer replied:
refactors gone wild :(

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/Unset.scala:
```diff
@@ -0,0 +1,21 @@
+  * cause all empty values to not be treated as deletes on write.
+  */
+case class CassandraOption[A](option: Option[A], unsetIfNone: Boolean = true) {
+  def UnsetIfNone = copy(option, true)
```
Perhaps such methods should be in a companion object

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/BoundStatementBuilder.scala:
```diff
@@ -38,11 +39,18 @@ private[connector] class BoundStatementBuilder[T](
+      //C* 2.2 (PV4) and Greater Allows us to leave fields Unset
+      if (protocolVersion.toInt < ProtocolVersion.V4.toInt || columnValue != Unset) {
+        if (columnValue == Unset){
+          logWarning(s"""Unset Value set to Null because Current $protocolVersion <
```
Please remove log message from there. This is executed for every single value which is inserted and will surely lead to kill the server (log file can grow much faster than the database!). Instead, you can make some flag and report warning once the whole batch is processed. Please do not log anything per column or even per row.

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/BoundStatementBuilder.scala:
```diff
@@ -38,11 +39,18 @@ private[connector] class BoundStatementBuilder[T](
-      if (serializedValue != null)
-        bytesCount += serializedValue.remaining()
+      //C* 2.2 (PV4) and Greater Allows us to leave fields Unset
+      if (protocolVersion.toInt < ProtocolVersion.V4.toInt || columnValue != Unset) {
```
I think we can make a function reference and assign one implementation or another once `BoundStatementBuilder` is created, and do not repeat here.

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

RussellSpitzer commented:
@jacek-lewandowski Comments addressed

RussellSpitzer commented:
I also added a writeConf parameter for use with SparkSQL

ds-jenkins-builds commented:
Build against Scala 2.11 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.10 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.10 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.11 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.11 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.10 finished with failure

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

jacek-lewandowski commented on doc/5_saving.md:
````diff
@@ -189,6 +189,135 @@ val company = Company("DataStax", address)
 sc.parallelize(Seq(company)).saveToCassandra("test", "companies")
 ```
 
+## Skipping Columns and Avoiding Tombstones on Writes (Connector Version 1.6+ and Cassandra 2.2+)
````
@RussellSpitzer can you apply code highlighting for code? Just enclose those snippets with \`\`\`scala ... \`\`\` or \`\`\`sql ... \`\`\`

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/TypeConverter.scala:
```diff
@@ -750,9 +768,19 @@ object TypeConverter {
 
     def targetTypeTag = implicitly[TypeTag[AnyRef]]
 
+    def cassandraOptionToAnyRef(cassandraOption: CassandraOption[_]) = {
```
Hmm... I'm still not convinced for this. This looks like messing different levels of encapsulation:
```scala
cassandraOption match {
  case CassandraOption.Value(x) => CassandraOption.Value(nestedConverter.convert(x).asInstanceOf[AnyRef])
  case x: CassandraOption => x
}
```
and make the writer take care of how it is handled.

Disregard this comment if you think it is a load of nonsense :)

RussellSpitzer replied:
Hmmm, So my feeling here was that it is the converter's responsibility to take in Scala Objects and output something that the Driver can actually insert. I think it's ok to remove the Unset => Unset (and Unset Object) but the Null => java `null` seems to be important to me.

Basically I was trying to make sure we were as close to as possible the OptionToNull converter (prior to these mods)

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/Unset.scala:
```diff
@@ -0,0 +1,43 @@
+package com.datastax.spark.connector.types
```
I suppose the file should be named `CassandraOption.scala` rather than `Unset.scala`

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/BoundStatementBuilder.scala:
```diff
@@ -38,12 +68,14 @@ private[connector] class BoundStatementBuilder[T](
       val converter = converters(i)
       val columnName = columnNames(i)
       val columnValue = converter.convert(buffer(i))
-      boundStatement.set(columnName, columnValue, CodecRegistryUtil.codecFor(columnTypes(i), columnValue))
```
Although I had something else on my mind when I commented this part of code for the last time, I'm ok with this solution.

My idea was more or else to create two methods with identical signature:
```scala
def bindColumnLegacy(stmt: RichBoundStatement, idx: Int, value: AnyRef) ...
```
```scala
def bindColumnV4(stmt: RichBoundStatement, idx: Int, value: AnyRef) ...
```
Both of them would include the appropriate implementation which would replace the code block below:
```scala
if (columnValue == Unset || (ignoreNulls && columnValue == null)) {
  maybeLeaveUnset(boundStatement, columnName)
} else {
  val codec = CodecRegistryUtil.codecFor(columnTypes(i), columnValue)
  boundStatement.set(columnName, columnValue, codec)
}
```
with appropriately inlined `maybeLeaveUnset`.

Then, you would have to do:
```scala
val bindColumn: (RichBoundStatement, Int, AnyRef) => Unit = protocolVersion match {
  case pv if pv.toInt <= ProtocolVersion.V3.toInt => bindColumnLegacy
  case _ => bindColumnV4
}
```
somewhere in main class block.

Ultimately, the it would look like:
```scala
    rowWriter.readColumnValues(row, buffer)
    var bytesCount = 0
    for (i <- 0 until columnNames.size) {
      val converter = converters(i)
      val columnName = columnNames(i)
      val columnValue = converter.convert(buffer(i))
      bindColumn(boundStatement, i, columnValue)
      val serializedValue = boundStatement.getBytesUnsafe(i)
      if (serializedValue != null) bytesCount += serializedValue.remaining()
    }
```
where `bindColumn` method would be preselected on whole `BoundStatementBuilder` creation. Semantically it would be identical to creation of two subclasses of `BoundStatementBuilder` - each with different definition of that `bindColumn` method.

RussellSpitzer replied:
Hmm I like this approach too. I wanted to make it clear that what portion of the build was different. I have no strong feelings either way I can switch it to this if you like? I'm not sure it makes much of a different performance wise since Hotspot should optimize this loop regardless of how we write it (i think)

jacek-lewandowski commented:
Apart from code highlighting and renaming one file I'm +1. Those other two comments are rather for discussion and I do not need them addressed there.

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

RussellSpitzer commented:
@jacek-lewandowski Addressed all comments. I redid the bindValue function like you described, if you could take a quick peak I think thats it for this ticket

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success

jacek-lewandowski commented:
+1, please squash before merging

RussellSpitzer commented:
Thanks will do

On Mon, Feb 1, 2016, 10:13 PM Jacek Lewandowski <notifications@github.com>
wrote:

> +1, please squash before merging
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/datastax/spark-cassandra-connector/pull/892#issuecomment-178395178>
> .
>

ds-jenkins-builds commented:
Build against Scala 2.11 finished with success

ds-jenkins-builds commented:
Build against Scala 2.10 finished with success
